### PR TITLE
Enable Live Markdown Preview on web

### PR DESCRIPTION
**Pull request by Skalakid**
> [!IMPORTANT]  
> QA team - please test Android & iOS as well since this PR also includes the upgrade of the package itself

### Details

This PR adds Live Markdown Preview feature for web using @expensify/react-native-live-markdown library. 
### Fixed Issues

$ https://github.com/Expensify/App/issues/27977
PROPOSAL:

### Tests

Testing Replacement Rules

Code Fence:
Type: ```your text```
Expected: Text appears in a separate block formatted as code.

Inline Code Block:
Type: ``your text``
Expected: Text appears inline, formatted as code.

Email Link:
Type: test@email.com
Expected: Email address turns into not clickable mailto link.

URL Link:
Type: `https://github.com/Expensify/App/pull/34292`
Expected: Displays `https://github.com/Expensify/App/pull/34292` styled as web link (it shouldn't be clickable).

Italics
Type: _your text_
Expected: Text appears in italics.

Bold Text:
Type: *your text*
Expected: Text appears in bold.

Strikethrough Text:
Type: ~your text~
Expected: Text appears with a strikethrough.

Block Quote:
Type: > your text
Expected: Text appears as a block quote.

Heading:
Type: # Your Heading
Expected: Text appears as a large, bold heading.

- [X] Verify that no errors appear in the JS console

### Offline tests

Same as Tests
### QA Steps

Same as Tests
- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [X] I verified that all the inputs inside a form are aligned with each other.
    - [X] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Uploading live-markdown-for-web.mov…

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/components/Composer/index.tsx
Comment (tomekzaw) [inline src/components/Composer/index.tsx:1]: Let's use `RNMarkdownTextInput` component instead. https://github.com/Expensify/App/blob/main/src/components/RNMarkdownTextInput.tsx

**Comment by thienlnam** [comment]
Adding the original reviewers of the live markdown web PR that we reverted

**Comment by c3024** [comment]
>Email Link:
Type: [test@email.com](mailto:test@email.com)
Expected: Email address turns into a clickable mailto link.

Should the email be clickable in the composer?

Both email and web links are not clickable in composer.

https://github.com/Expensify/App/assets/102477862/d8839fe4-506a-425f-af72-ba5dbab6a45d

**Comment by c3024** [comment]
A warning is thrown after logging in

Screenshot 2024-03-14 at 8 58 22 AM

**Comment by c3024** [comment]
3 level indents in composer for quote against 4 levels in the sent message.

https://github.com/Expensify/App/assets/102477862/5ff799a1-8b65-4c31-9f8e-eaea07d05a62

**Comment by BartoszGrajdek** [comment]
> A warning is thrown after logging in

On it! 👀

**Comment by BartoszGrajdek** [comment]
> A warning is thrown after logging in

Should be ok now 😄

**Comment by thienlnam** [comment]
cc @c3024 / @shubham1206agra Generated a new build so we can keep testing cycles fast

**Comment by c3024** [comment]
Still seeing the `flushSync` error. 🤔 
Screenshot 2024-03-15 at 8 44 08 PM

**Comment by c3024** [comment]
When the text is multiline the markdown for the link followed is lost. It gets parsed fine after sending the message.


https://github.com/Expensify/App/assets/102477862/d18039a5-23fe-4336-80d3-5d9a0dae9ac2

**Comment by c3024** [comment]
Need to click Send twice to send the message when I write some text and paste an image in the composer. On prod, clicking send once sends the message.


https://github.com/Expensify/App/assets/102477862/b38ab879-7be8-4ccf-bc95-72d3a7104f1e

**Comment by c3024** [comment]
Also please merge main and fix conflicts.

**Comment by c3024** [comment]
Pasting in the editing composer pastes the text twice in main composer.


https://github.com/Expensify/App/assets/102477862/29c68ab6-c190-4135-a511-bd7e15d911bd


https://github.com/Expensify/App/assets/102477862/d5f218df-b420-421f-a7a4-eb309fd1769a

**Comment by BartoszGrajdek** [comment]
Working on fixes today! 😄

**Comment by BartoszGrajdek** [comment]
Alright, I managed to solve problems with flushSync error & pasting in the wrong composer. 

I'll continue working on solving the other pasting issues tomorrow - they are likely related to each other, but the solution itself is a little more advanced.

**Comment by BartoszGrajdek** [comment]
> When the text is multiline the markdown for the link followed is lost. It gets parsed fine after sending the message.

It's working the same way on Android implementation, so if we decide to fix it I think it's better to do it in a separate issue.

> I am using this text for testing preview on long text. This takes a lot of time to paste right now.

As per Jack mentioned in [slack](https://swmansion.slack.com/archives/C06BDSWLDPB/p1710803516654419?thread_ts=1710803472.802019&cid=C06BDSWLDPB) this is not something we should consider a blocker for now

The duplicate pasting issue (& one with attachments) should be fixed now, but it required disabling a utility function created by Robert from our team at SWM, so I'll discuss it with him tomorrow and make sure it didn't break anything before we test that part once again 😄

**Comment by BartoszGrajdek** [comment]
Alright I've discussed it with Robert and there shouldn't be any problems with this solution, so this PR is ready for testing now! 😄

**Comment by thienlnam** [comment]
cc @c3024 / @shubham1206agra Could we do another review on this? Seems like we're pretty close here

File: src/pages/home/report/ReportActionCompose/SuggestionMention.tsx
Comment (Kicu) [inline src/pages/home/report/ReportActionCompose/SuggestionMention.tsx:289]: Some notes for this one: 
 - removing this solves the bug when user would delete a character on the mention text and the mentions suggestion list would not appear
 - originally it was introduced here: https://github.com/Expensify/App/pull/29957 (solving issue: https://github.com/Expensify/App/issues/28170)
 - this was described as a bug in `TextInput` but now on this branch `RNMarkdownTextInput` is used and this bug no longer appears
 - I have verified on mac os Chrome, Ios, Android and mWeb ios that without this if we do not experience any kind of mentions flashing when user deletes a character immediately before the `@` char

**Comment by c3024** [comment]
Ref: https://github.com/Expensify/App/pull/38152#issuecomment-2000119081

>Need to click Send twice to send the message when I write some text and paste an image in the composer. On prod, clicking send once sends the message.

Now, it needs clicking `Send` three times on Dev. 🤔 💭 


https://github.com/Expensify/App/assets/102477862/32636a8d-6d28-4884-a339-5e60373a2f8a


https://github.com/Expensify/App/assets/102477862/a909a4f0-bda8-4b80-8a78-799d112a6c11

**Comment by BartoszGrajdek** [comment]
> Now, it needs clicking Send three times on Dev. 🤔 💭

That's weird - does it always happen to you, or does a refresh / hard refresh fix it?

**Comment by c3024** [comment]
> > Now, it needs clicking Send three times on Dev. 🤔 💭
> 
> That's weird - does it always happen to you, or does a refresh / hard refresh fix it?

Thanks, it did work.

**Comment by BartoszGrajdek** [comment]
> Thanks, it did work.

Still, we should fix it - I'll try working on it today! 😄

**Comment by BartoszGrajdek** [comment]
Quick update: I have a solution for the pasting issue, but I need to test if it doesn't create any regressions, so I'll be handling that on Monday 🙌🏻

**Comment by BartoszGrajdek** [comment]
This should fix the pasting issue 😄

**Comment by BartoszGrajdek** [comment]
@shubham1206agra not sure if you have access to [this Slack thread](https://swmansion.slack.com/archives/C06BDSWLDPB/p1711031417895639?thread_ts=1710803472.802019&cid=C06BDSWLDPB), so I'll summarize it here:

During the preparation for this PR we noticed that the library is prone to errors related with scrolling/cursor movement and already fixed a couple of these, but we realised that there's likely a better solution for all of this that should prevent similar problems in the future.

A root cause for these bugs is how we're handling text & linebreaks, that's why [this issue](https://github.com/Expensify/react-native-live-markdown/issues/239) was created. This is something we will be working on after we get this PR into staging.

Jack has agreed, that this is something we can continue working on later

**Comment by thienlnam** [comment]
Thanks for catching these - I think the extra line break can also be addressed after we merge this PR. I generated a new adhoc - but it sounds like we've solved the main ones that would be blockers

**Comment by tomekzaw** [comment]
Thanks @thienlnam for generating adhoc builds. This PR bumps `@expensify/react-native-live-markdown` from `0.1.5` to a higher version so it also affects the native app; we might want to QA the message composer on Android and iOS as well.

**Comment by c3024** [comment]
When I paste large text with a mention in it the cursor gets placed inside the mention instead of the end of the text.


https://github.com/Expensify/App/assets/102477862/9a4e0b66-85b7-4723-9b68-fb6ebc957475

**Comment by c3024** [comment]
There is unused space in the composer when expanding the composer.


https://github.com/Expensify/App/assets/102477862/57fddc32-4de1-45d0-ab57-91e75c48bde2

**Comment by BartoszGrajdek** [comment]
> There is unused space in the composer when expanding the composer.

Can you explain what you mean?

**Comment by c3024** [comment]
The scroll does not extend throughout the composer height when clicked on the expand icon on the top left of the composer. It works fine on `main`.

https://github.com/Expensify/App/assets/102477862/1f09513a-10cd-4da8-8ddb-a0dcd3bde479

**Comment by BartoszGrajdek** [comment]
> The scroll does not extend throughout the composer height when clicked on the expand icon on the top left of the composer. It works fine on main.

Should be fine now

**Comment by thienlnam** [comment]
Okay generating one last adhoc for those new changes - let's merge this if we don't find any blockers.

I've been adding these other edge cases to these issues that aren't blockers https://github.com/Expensify/App/issues/39006

**Comment by Skalakid** [comment]
> BUG: Undo operation removes more than one character

@shubham1206agra I think it's not a bug. History items are added with the debouce, so if you are writing faster, parts of the text are saved and then undone

**Comment by BartoszGrajdek** [comment]
> For mentions, if the mention is followed by a space then scrolls works correctly on pressing Shift + Enter but if we press Shift + Enter right after the mention without a space followed then it does not scroll.

Also fixed 🚀

**Review by thienlnam — APPROVED** [review]
This has gone through a series of reviews and we've fixed any main blockers - I tested one final time and everything looks good - let's ship this out and fix the follow ups as they come up

**Comment by thienlnam** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by BartoszGrajdek** [comment]
> [!IMPORTANT]
> QA team - please test Android & iOS as well since this PR also includes the upgrade of the package itself

**Pull request merged**
